project1/server/server.cpp

```diff
@@ -10,105 +10,32 @@
 
 #include "agent.h"
 #include "file_handler/file_handler.h"
+#include "../thread_pool/thread_pool.h"
+#include "server_tasks/nport_task.h"
+#include "server_tasks/tport_task.h"
 
 namespace server {
-namespace {
 
-using file_handler::FileHandler;
+    void Server::FtpService(uint16_t nPort, uint16_t tPort) {
 
-/// Maximum queue size to use for listening on the server socket.
-constexpr uint8_t kMaxQueueSize = 5;
+        thread_pool::ThreadPool pool;
 
-/**
- * @brief Helper function that creates the address structure to use for the
- *   server.
- * @param port The port to listen on.
- * @return The address structure.
- */
-struct sockaddr_in MakeAddress(uint16_t port) {
-  struct sockaddr_in address {};
-  address.sin_family = AF_INET;
-  address.sin_addr.s_addr = INADDR_ANY;
-  address.sin_port = htons(port);
+        std::shared_ptr<server_tasks::CommandIDs> active_ids;
+        auto nPortTask = std::make_shared<server_tasks::NPortTask>();
+        auto tPortTask = std::make_shared<server_tasks::TPortTask>();
 
-  return address;
-}
+        // pass active command list to nPortTask and tPortTask
+        nPortTask->SetActiveCommandIDs(active_ids);
+        tPortTask->SetActiveCommandIDs(active_ids);
 
-/**
- * @brief Sets up the server socket. After this is complete, the socket is
- * ready to accept new connections.
- * @param The address structure to use for configuring the socket.
- * @return The socket FD if successful, or -1 if not.
- */
-int SetUpSocket(const struct sockaddr_in &address) {
-  // Open a TCP socket.
-  const int server_fd = socket(AF_INET, SOCK_STREAM, 0);
-  if (server_fd == 0) {
-    perror("Failed to create server socket");
-    return -1;
-  }
+        // pass file access managers to nPortTask
+        nPortTask->SetFileAccessManagers(read_manager_,write_manager_);
 
-  // Allow the server to re-bind to this port if it was restarted quickly.
-  const int option = 1;
-  if (setsockopt(server_fd, SOL_SOCKET, SO_REUSEADDR | SO_REUSEPORT, &option,
-                 sizeof(option))) {
-    perror("Failed to set socket options");
-    // This is not a fatal error.
-  }
+        nPortTask->SetPort(nPort);
+        tPortTask->SetPort(tPort);
 
-  // Bind to the port.
-  if (bind(server_fd, (struct sockaddr *)&address, sizeof(address)) < 0) {
-    perror("bind() failed on server socket");
-    return -1;
-  }
-  if (listen(server_fd, kMaxQueueSize) < 0) {
-    perror("listen() failed on server socket");
-    return -1;
-  }
+        pool.AddTask(nPortTask);
+        pool.AddTask(tPortTask);
```

Comment:
So, as you have it now, when control gets to here, it's going to destroy the pool and the tasks you just added will be immediately cancelled. You need some way of keeping the main thread alive indefinitely. I suggest you sit here and wait for all the tasks in the pool to exit naturally.

Reply:
How would I go about waiting all tasks to exit naturally? I see some of the examples in your tests but I'm not sure how that would apply here

Reply:
The trivial way is just checking the statuses of the two tasks you added in a loop every second or so and exiting when they both flip to something other than `RUNNING`. I'm also working on a better way to do this.